### DISCUSSION TITLE
"Get an ApplicationSecret instance by some means" for google-youtube3 crate

### DISCUSSION THREAD

**Discussion by randall-coding** [Byron/google-apis-rs#330] [Q&A] [answered]
I was just wondering if you could provide more clarification on the state of the google-youtube3 code example in the readme. 

**"Get an ApplicationSecret instance by some means"**

I'm new to rust and the youtube3 api and I'm not sure how I'd go about obtaining this instance in order to authenticate.  Any help is greatly appreciated.

**Comment by Byron** [comment, marked as answer]
These crates probably assume prior experience with the usage of Google APIs, maybe that's something to improve on indeed.

One will have to use the google developer console to associate APIs with an application and then download an application secret JSON file. That's roughly the steps to follow and should be enough to go on and I hope it helps.

You are welcome to contribute adjustments to the documentation, there probably is a google manual to link to as minimal but significant improvement.

**Reply by randall-coding** [reply]
Ah so that authorization is in regards to associating my application with Google's API.  So from there I guess I would still need to gather the user's credentials and authorize them.  Do you know if it is possible for me to use pre-stored username/password to authenticate them rather than taking them through the SSO portal?

**Reply by Byron** [reply]
> Do you know if it is possible for me to use pre-stored username/password to authenticate them rather than taking them through the SSO portal?

Yes, this is what the `yup-oauth2` crate allows to do. Note that a choice is only provided when using the APIs as libraries, not when using their CLIs.

**Reply by randall-coding** [reply]
Thanks.

**Reply by CodyBontecou** [reply]
Hey Byron, I have my Google API's application secret JSON file. In a previous application I built, this file's path was provided through an environment variable and the rest was taken care of.

Being new to Rust, I'm unsure how to attach this json file to the secret variable in a way it will work.

**Reply by Byron** [reply]
> Being new to Rust, [..]

This isn't related to how Rust works and entirely up to the crate implementation. In this case, when running the CLI it will tell about the path it tries to read the secrets from. [Here is the code](https://github.com/Byron/google-apis-rs/blob/main/src/mako/cli/lib/engine.mako#L106:L129) that the CLI uses to handle secrets and obtain tokens. From there it will be able to pick them up automatically during next invocation or refresh them if needed. I hope that helps.
